url79d4f7c7ed_schanq_Live_Algorithm_tgz-pJ8-algorithms_and_systems_AI_Evolutionary_GeneticAlgorithmJ8/src/algorithms_and_systems_AI_Evolutionary/GeneticAlgorithm.java

```diff
@@ -0,0 +1,504 @@
+// ##################################################################
+package algorithms_and_systems_AI_Evolutionary;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Random;
+
+import expert_system.DesiredLoudness;
+import expert_system.Dynamic;
+import expert_system.DynamicShape;
+import expert_system.Feel;
+import expert_system.Key;
+import expert_system.Rhythmic_Block;
+import expert_system.Scale;
+import expert_system.Tonality;
+
+
+public class GeneticAlgorithm {
+
+	// ##########################################################
+	// #					     FIELDS							#
+	// ##########################################################
+	
+	public Tonality 		 targetTonality;
+	public DesiredLoudness 	 targetLoudness;
+	public Feel				 targetFeel;
+	
+	public int 				 populationSize = 50;
+	public int 				 totalFitnessOfPopulation = 0;
+	public int 				 theFittest = 0;
+	public int 				 minOctave, maxOctave;
+	
+	public List <Chromosome> citezens;
+	
+	public Random 	 		 random;
+
+	// ##########################################################
+	// #					   CONSTRUCTORS						#
+	// ##########################################################
+	
+	public GeneticAlgorithm(TargetChromosome t, int min, int max){
+		
+		this.random 		= new Random();
+		
+		this.targetTonality = t.getTargetTonality();
+		this.targetLoudness = t.getTargetLoudness();
+		this.targetFeel		= t.getTargetFeel();
+		this.minOctave 		= min;
+		this.maxOctave 		= max;
+		this.citezens 		= new ArrayList <Chromosome> ();
+		
+		
+	}
+	// ##########################################################
+	// #	METHODS: for the initialisation of Population,  	#
+	// #			 calculation of fitness, selection of 		#
+	// #			 Parents and reproducion					#
+	// ##########################################################
+	// ==========================================================
+	// Returns populationSize field								=
+	// ==========================================================
+	public int getPopulationSize() {
+		return populationSize;
+	}
+	// ==========================================================
+	// Sets the populationSize field to the value of the input	=
+	// parameter 												=
+	// ==========================================================
+	public void setPopulationSize(int populationSize) {
+		this.populationSize = populationSize;
+	}
+	
+	// Initialises the citezen list with a population of 
+	// chromosomes with a set of random genes
+	public void populate(){
+		
+		for(int i = 0; i < this.populationSize; i++){
+			Chromosome chr = new Chromosome(this.minOctave, this.maxOctave);
+			this.citezens.add(chr);
+		}
+	}
+	// ==========================================================
+	// Uses a fitness function to determine how fit the 		=
+	// input Chromosome is in comparison to the target			=
+	// ==========================================================
+	public void calculateFitness(Chromosome c){
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 14. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks on lines 100-113, lines 117-129, lines 138-156 into separate methods.